### PR TITLE
feat: add contextual tuples to Expand API

### PR DESCRIPTION
**Pull request by sujitha-av**
## Description
Adds contextual tuples support to Expand request.

## References

#951 

1. [PR](https://github.com/openfga/api/pull/202) to add Contextual tuples to the request in openfga/api
2. [PR](https://github.com/openfga/api/pull/209) to add documentation for the same

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above] - Yet to
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Review by miparnisari — COMMENTED** [review]
Thanks @sujitha-av. Could you run `go get github.com/openfga/api@sujitha-av:contextual-tuples-expand-api` and update the go.mod? I want to make sure that everything compiles.

File: pkg/server/commands/expand.go [hunk 1]
```diff
@@ -55,25 +57,19 @@ func (q *ExpandQuery) Execute(ctx context.Context, req *openfgav1.ExpandRequest)
...
-		return nil, serverErrors.HandleError("", err)
+	typesys, ok := typesystem.TypesystemFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("%w: typesystem missing in context", openfgaErrors.ErrUnknown)
 	}
 
-	if !typesystem.IsSchemaVersionSupported(model.GetSchemaVersion()) {
-		return nil, serverErrors.ValidationError(typesystem.ErrInvalidSchemaVersion)
+	for _, ctxTuple := range req.GetContextualTuples().GetTupleKeys() {
+		if err := validation.ValidateTuple(typesys, ctxTuple); err != nil {
```
Comment (miparnisari) [inline pkg/server/commands/expand.go:66]:
```suggestion
		if err := validation.ValidateTupleForWrite(typesys, ctxTuple); err != nil {
```

File: pkg/server/commands/expand_test.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+			ds := memory.New()
+			t.Cleanup(ds.Close)
+			storeID, model := storagetest.BootstrapFGAStore(t, ds, tc.modelStr, tc.tuples)
+			ts, err := typesystem.NewAndValidate(
+				context.Background(),
+				model,
+			)
+			require.NoError(t, err)
+
+			ctx := storage.ContextWithRelationshipTupleReader(context.Background(), ds)
```
Comment (miparnisari) [inline pkg/server/commands/expand_test.go:157]: Tests pass without this line, right? if so, please remove it

**Comment by sujitha-av** [comment]
> run go get github.com/openfga/api@sujitha-av:contextual-tuples-expand-api and update the go.mod

Hi. I am not able to do this and not sure if I should commit the changes in go.mod too?

**Comment by miparnisari** [comment]
@sujitha-av yes please update the go.mod, if the code doesn't compile the build will not pass and we won't be able to see whether the PR is good to be merged.

**Review by miparnisari — COMMENTED** [review]
Also please update CHANGELOG.md :)

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,7 @@ Try to keep listed changes to a concise bulleted list of simple explanations of
 
 ### Added
 * Added `start_time` parameter to `ReadChanges` API to allow filtering by specific time [#2020](https://github.com/openfga/openfga/pull/2020)
+* Added support for Contextual Tuples in the `Expand` API.
```
Comment (miparnisari) [inline CHANGELOG.md:13]:
```suggestion
* Added support for Contextual Tuples in the `Expand` API. [#2045](https://github.com/openfga/openfga/pull/2045)
```

File: pkg/server/server.go [hunk 4]
```diff
@@ -1220,11 +1220,14 @@ func (s *Server) Expand(ctx context.Context, req *openfgav1.ExpandRequest) (*ope
...
-	return q.Execute(ctx, &openfgav1.ExpandRequest{
-		StoreId:              storeID,
-		AuthorizationModelId: typesys.GetAuthorizationModelID(), // the resolved model id
-		TupleKey:             tk,
-		Consistency:          req.GetConsistency(),
+	return q.Execute(
+		typesystem.ContextWithTypesystem(ctx, typesys), 
+		&openfgav1.ExpandRequest{
+			StoreId:              storeID,
+			AuthorizationModelId: typesys.GetAuthorizationModelID(), // the resolved model id
```
Comment (miparnisari) [inline pkg/server/server.go:1227]:
```suggestion
			AuthorizationModelId: typesys.GetAuthorizationModelID(), // the resolved model id
```

do we need this line?

Comment (sujitha-av) [reply pkg/server/server.go:1227]: yes, it is not necessary, removed.

**Comment by sujitha-av** [comment]
Can I know why security/snyk (OpenFGA) and license/snyk (OpenFGA) workflows have failed and what I can do?

File: pkg/server/commands/expand_test.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	"github.com/stretchr/testify/require"
+
+	serverErrors "github.com/openfga/openfga/pkg/server/errors"
+	"github.com/openfga/openfga/pkg/storage/memory"
+	storagetest "github.com/openfga/openfga/pkg/storage/test"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+func TestExpandWithContextualTuples(t *testing.T) {
```
Comment (miparnisari) [inline pkg/server/commands/expand_test.go:17]:
```suggestion
func TestExpand(t *testing.T) {
```

**Review by miparnisari — APPROVED** [review]

**Comment by miparnisari** [comment]
> Can I know why security/snyk (OpenFGA) and license/snyk (OpenFGA) workflows have failed and what I can do?

It's because of this line `replace github.com/openfga/api/proto => github.com/sujitha-av/api/proto v0.0.0-20241031191721-50d1c0a9bae3`

we need to merge the API PR and then bump this one.

File: go.mod [hunk 6]
```diff
@@ -134,3 +134,5 @@ require (
 	modernc.org/strutil v1.2.0 // indirect
 	modernc.org/token v1.1.0 // indirect
 )
+
+replace github.com/openfga/api/proto => github.com/sujitha-av/api/proto v0.0.0-20241031191721-50d1c0a9bae3
```
Comment (miparnisari) [inline go.mod:138]:
```suggestion
```

File: go.sum [hunk 7]
```diff
@@ -243,6 +241,8 @@ github.com/stretchr/testify v1.9.0 h1:HtqpIVDClZ4nwg75+f6Lvsy/wHu+3BoSGCbBAcpTsT
 github.com/stretchr/testify v1.9.0/go.mod h1:r2ic/lqez/lEtzL7wO/rwa5dbSLXVDPFyf8C91i36aY=
 github.com/subosito/gotenv v1.6.0 h1:9NlTDc1FTs4qu0DDq7AEtTPNw6SVm7uBMsUCUjABIf8=
 github.com/subosito/gotenv v1.6.0/go.mod h1:Dk4QP5c2W3ibzajGcXpNraDfq2IrhjMIvMSWPKKo0FU=
+github.com/sujitha-av/api/proto v0.0.0-20241031191721-50d1c0a9bae3 h1:dZ1WNxPTwFRtcDRyXl0YVESyh5txVTxpGRDf6XMOitc=
+github.com/sujitha-av/api/proto v0.0.0-20241031191721-50d1c0a9bae3/go.mod h1:gil5LBD8tSdFQbUkCQdnXsoeU9kDJdJgbGdHkgJfcd0=
```
Comment (miparnisari) [inline go.sum:245]:
```suggestion
github.com/openfga/api/proto v0.0.0-20241104193559-ee46d6721514 h1:ebCcwMV9LTbAG7so6PQt5qCLB6dz5VUW7/pf9sokG5M=
github.com/openfga/api/proto v0.0.0-20241104193559-ee46d6721514/go.mod h1:gil5LBD8tSdFQbUkCQdnXsoeU9kDJdJgbGdHkgJfcd0=
```

**Review by willvedd — APPROVED** [review]

**Pull request merged**